File: app/integrations/github_client.py
"""Клиент GitHub REST API для одностороннего импорта issues."""

from dataclasses import dataclass, field
from typing import Protocol

import httpx

from app.services.exceptions import BadRequestError, NotFoundError, ServiceUnavailableError

_PER_PAGE = 100


@dataclass(frozen=True)
class GitHubIssue:
    number: int
    title: str
    body: str | None
    state: str
    assignee_login: str | None
    labels: list[str] = field(default_factory=list)


class GitHubIssueSource(Protocol):
    async def fetch_issues(self) -> list[GitHubIssue]: ...


def _parse_issue(payload: dict) -> GitHubIssue:
    assignee = payload.get("assignee")
    return GitHubIssue(
        number=payload["number"],
        title=payload["title"],
        body=payload.get("body"),
        state=payload["state"],
        assignee_login=assignee["login"] if assignee else None,
        labels=[label["name"] for label in payload.get("labels", [])],
    )


class GitHubClient:
    """Читает issues репозитория через GitHub REST API.

    transport инжектируется в тестах (httpx.MockTransport), сеть не нужна.
    """

    def __init__(
        self,
        repo: str,
        api_url: str,
        token: str | None = None,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        self.repo = repo
        self.api_url = api_url.rstrip("/")
        self.token = token
        self.transport = transport

    def _headers(self) -> dict[str, str]:
        headers = {"Accept": "application/vnd.github+json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        return headers

    def _raise_for_status(self, response: httpx.Response) -> None:
        if response.status_code == 404:
            raise NotFoundError(f"GitHub repository {self.repo} not found")
        if response.status_code in (401, 403):
            raise BadRequestError("GitHub rejected the request: check token and rate limits")
        if response.status_code == 429 or response.status_code >= 500:
            raise ServiceUnavailableError("GitHub is temporarily unavailable, try again later")
        if response.status_code >= 400:
            raise BadRequestError(f"GitHub API error: HTTP {response.status_code}")

    async def fetch_issues(self) -> list[GitHubIssue]:
        issues: list[GitHubIssue] = []
        try:
            async with httpx.AsyncClient(transport=self.transport) as client:
                page = 1
                while True:
                    response = await client.get(
                        f"{self.api_url}/repos/{self.repo}/issues",
                        params={"state": "all", "per_page": _PER_PAGE, "page": page},
                        headers=self._headers(),
                    )
                    self._raise_for_status(response)
                    payload = response.json()
                    issues.extend(
                        # Endpoint отдаёт и pull requests — их отличает ключ pull_request.
                        _parse_issue(item)
                        for item in payload
                        if "pull_request" not in item
                    )
                    if len(payload) < _PER_PAGE:
                        return issues
                    page += 1
        except httpx.HTTPError as exc:
            raise ServiceUnavailableError(
                "GitHub is unreachable: check network and proxy settings"
            ) from exc

File: backend/src/main/java/com/skilltracker/integration/GitHubClient.java
package com.skilltracker.integration;

import com.skilltracker.config.GitHubProperties;
import com.skilltracker.exception.BadRequestException;
import com.skilltracker.exception.NotFoundException;
import com.skilltracker.exception.ServiceUnavailableException;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.ResourceAccessException;
import org.springframework.web.client.RestClient;
import tools.jackson.databind.JsonNode;

/** Reads issues of the configured repository through the GitHub REST API. */
@Component
public class GitHubClient implements GitHubIssueSource {

    private static final int PER_PAGE = 100;
    private static final String ACCEPT = "application/vnd.github+json";

    private final RestClient restClient;
    private final GitHubProperties properties;

    public GitHubClient(RestClient.Builder restClientBuilder, GitHubProperties properties) {
        this.restClient = restClientBuilder.build();
        this.properties = properties;
    }

    @Override
    public List<GitHubIssue> fetchIssues() {
        List<GitHubIssue> issues = new ArrayList<>();

        int page = 1;
        while (true) {
            JsonNode payload = requestPage(page);
            int received = payload.isArray() ? payload.size() : 0;

            for (JsonNode item : payload) {
                // The endpoint also returns pull requests; they carry a pull_request key.
                if (!item.has("pull_request")) {
                    issues.add(parseIssue(item));
                }
            }

            if (received < PER_PAGE) {
                return issues;
            }
            page++;
        }
    }

    private JsonNode requestPage(int page) {
        ResponseEntity<JsonNode> response;
        try {
            response = restClient
                    .get()
                    .uri(issuesUri(page))
                    .accept(MediaType.parseMediaType(ACCEPT))
                    .headers(this::applyAuthorization)
                    .retrieve()
                    .onStatus(HttpStatusCode::isError, (request, errorResponse) -> {})
                    .toEntity(JsonNode.class);
        } catch (ResourceAccessException unreachable) {
            throw new ServiceUnavailableException("GitHub is unreachable: check network and proxy settings");
        }

        raiseForStatus(response.getStatusCode().value());
        JsonNode body = response.getBody();
        return body == null ? tools.jackson.databind.node.JsonNodeFactory.instance.arrayNode() : body;
    }

    /**
     * Built as a whole URI rather than through a template: the { owner/name} slug carries a
     * slash that template expansion would percent-encode into a path segment GitHub does not know.
     */
    private URI issuesUri(int page) {
        return URI.create(baseUrl() + "/repos/" + properties.repo() + "/issues" + "?state=all&per_page=" + PER_PAGE
                + "&page=" + page);
    }

    private void applyAuthorization(HttpHeaders headers) {
        if (properties.hasToken()) {
            headers.setBearerAuth(properties.token());
        }
    }

    private void raiseForStatus(int status) {
        if (status == 404) {
            throw new NotFoundException("GitHub repository " + properties.repo() + " not found");
        }
        if (status == 401 || status == 403) {
            throw new BadRequestException("GitHub rejected the request: check token and rate limits");
        }
        if (status == 429 || status >= 500) {
            throw new ServiceUnavailableException("GitHub is temporarily unavailable, try again later");
        }
        if (status >= 400) {
            throw new BadRequestException("GitHub API error: HTTP " + status);
        }
    }

    private String baseUrl() {
        String apiUrl = properties.apiUrl();
        return apiUrl.endsWith("/") ? apiUrl.substring(0, apiUrl.length() - 1) : apiUrl;
    }

    private GitHubIssue parseIssue(JsonNode payload) {
        JsonNode assignee = payload.get("assignee");
        List<String> labels = new ArrayList<>();
        JsonNode labelNodes = payload.get("labels");
        if (labelNodes != null && labelNodes.isArray()) {
            labelNodes.forEach(label -> labels.add(label.path("name").asString()));
        }

        return new GitHubIssue(
                payload.path("number").asInt(),
                payload.path("title").asString(),
                payload.hasNonNull("body") ? payload.get("body").asString() : null,
                payload.path("state").asString(),
                assignee != null && !assignee.isNull() ? assignee.path("login").asString() : null,
                labels);
    }
}
